Source/WebCore/rendering/RenderBox.cpp

```diff
@@ -2021,8 +2021,18 @@ bool RenderBox::pushContentsClip(PaintInfo& paintInfo, const LayoutPoint& accumu
     float deviceScaleFactor = document().deviceScaleFactor();
     FloatRect clipRect = snapRectToDevicePixels((isControlClip ? controlClipRect(accumulatedOffset) : overflowClipRect(accumulatedOffset, nullptr, IgnoreOverlayScrollbarSize, paintInfo.phase)), deviceScaleFactor);
     paintInfo.context().save();
-    if (style().hasBorderRadius())
-        paintInfo.context().clipRoundedRect(style().getRoundedInnerBorderFor(LayoutRect(accumulatedOffset, size())).pixelSnappedRoundedRectForPainting(deviceScaleFactor));
+    if (style().hasBorderRadius()) {
+        auto innerBorder = style().getRoundedInnerBorderFor(LayoutRect(accumulatedOffset, size()));
+        if (innerBorder.isRenderable())
+            paintInfo.context().clipRoundedRect(innerBorder.pixelSnappedRoundedRectForPainting(deviceScaleFactor));
+        else {
+            paintInfo.context().clipRoundedRect(calculateAdjustedInnerBorder(innerBorder, BoxSide::Top).pixelSnappedRoundedRectForPainting(deviceScaleFactor));
+            paintInfo.context().clipRoundedRect(calculateAdjustedInnerBorder(innerBorder, BoxSide::Bottom).pixelSnappedRoundedRectForPainting(deviceScaleFactor));
+            paintInfo.context().clipRoundedRect(calculateAdjustedInnerBorder(innerBorder, BoxSide::Left).pixelSnappedRoundedRectForPainting(deviceScaleFactor));
+            paintInfo.context().clipRoundedRect(calculateAdjustedInnerBorder(innerBorder, BoxSide::Right).pixelSnappedRoundedRectForPainting(deviceScaleFactor));
+        }
```

Comment:
Maybe you have a helper for this that takes GraphicsContext a parameter, clipContentForBorderRadius or something.

Reply:
It seems wrong to compute a clip for each side (ending up with the intersection of 4 clips) rather than just having logic that fixes up the rounded rect.

Reply:
Why not just do:
```
diff --git a/Source/WebCore/rendering/RenderBox.cpp b/Source/WebCore/rendering/RenderBox.cpp
index 704c4286a85d4d304024cf7043d8f82637b39c34..6e6b63fc0229d372080fe08ae647ea8dfa778f89 100644
--- a/Source/WebCore/rendering/RenderBox.cpp
+++ b/Source/WebCore/rendering/RenderBox.cpp
@@ -2021,8 +2021,13 @@ bool RenderBox::pushContentsClip(PaintInfo& paintInfo, const LayoutPoint& accumu
     float deviceScaleFactor = document().deviceScaleFactor();
     FloatRect clipRect = snapRectToDevicePixels((isControlClip ? controlClipRect(accumulatedOffset) : overflowClipRect(accumulatedOffset, nullptr, IgnoreOverlayScrollbarSize, paintInfo.phase)), deviceScaleFactor);
     paintInfo.context().save();
-    if (style().hasBorderRadius())
-        paintInfo.context().clipRoundedRect(style().getRoundedInnerBorderFor(LayoutRect(accumulatedOffset, size())).pixelSnappedRoundedRectForPainting(deviceScaleFactor));
+    if (style().hasBorderRadius()) {
+        auto innerRoundedRect = style().getRoundedInnerBorderFor(LayoutRect(accumulatedOffset, size())).pixelSnappedRoundedRectForPainting(deviceScaleFactor);
+        if (!innerRoundedRect.isRenderable())
+            innerRoundedRect.adjustRadii();
+        paintInfo.context().clipRoundedRect(innerRoundedRect);
+    }
+    
     paintInfo.context().clip(clipRect);
 
     if (paintInfo.phase == PaintPhase::EventRegion)
```

Reply:
Fixed.